src/applications/covid19-chatbot/index.js

```diff
@@ -0,0 +1,170 @@
+
+  // Before assigning, ensure it's a valid locale string (xx or xx-XX)
+  if (localeParam.search(localeRegExPattern) === 0) {
+    return localeParam;
+  }
+  return defaultLocale;
+}
+
+function getUserLocation(callback) {
+  navigator.geolocation.getCurrentPosition(
+    position => {
+      const latitude = position.coords.latitude;
+      const longitude = position.coords.longitude;
+      const location = {
+        lat: latitude,
+        long: longitude,
+      };
+      callback(location);
+    },
+    error => {
+      // user declined to share location
+      // eslint-disable-next-line no-console
+      console.log(`location error:${error.message}`);
+      callback();
+    },
+  );
+}
+
+function startChat(user, webchatOptions) {
+  window.WebChat.renderWebChat(webchatOptions, root);
+}
+
+function initBotConversation() {
+  if (this.status >= 400) {
+    alert(this.statusText);
+    return;
+  }
+  // extract the data from the JWT
+  const jsonWebToken = this.response;
+  const tokenPayload = JSON.parse(atob(jsonWebToken.split('.')[1]));
+  const user = {
+    id: tokenPayload.userId,
+    name: tokenPayload.userName,
+    locale: tokenPayload.locale,
+  };
+  let domain = undefined;
+  if (tokenPayload.directLineURI) {
+    domain = `https://${tokenPayload.directLineURI}/v3/directline`;
+  }
+  const botConnection = window.WebChat.createDirectLine({
+    token: tokenPayload.connectorToken,
+    domain,
+  });
+  const styleOptions = {
+    hideSendBox: true,
+    botAvatarInitials: 'VA',
+    userAvatarInitials: 'You',
+    backgroundColor: '#F8F8F8',
+    primaryFont: 'Source Sans Pro, sans-serif',
+  };
+
+  const webchatStore = window.WebChat.createStore(
+    {},
+    store => next => action => {
+      if (action.type === 'DIRECT_LINE/CONNECT_FULFILLED') {
+        store.dispatch({
+          type: 'DIRECT_LINE/POST_ACTIVITY',
+          meta: { method: 'keyboard' },
+          payload: {
+            activity: {
+              type: 'invoke',
+              name: 'InitConversation',
+              locale: user.locale,
+              value: {
+                // must use for authenticated conversation.
+                jsonWebToken,
+
+                // Use the following activity to proactively invoke a bot scenario
+
+                triggeredScenario: {
+                  trigger: chatBotScenario,
+                },
+              },
+            },
+          },
+        });
+      } else if (action.type === 'DIRECT_LINE/INCOMING_ACTIVITY') {
+        if (
+          action.payload &&
+          action.payload.activity &&
+          action.payload.activity.type === 'event' &&
+          action.payload.activity.name === 'ShareLocationEvent'
+        ) {
+          // share
+          getUserLocation(location => {
+            store.dispatch({
+              type: 'WEB_CHAT/SEND_POST_BACK',
+              payload: { value: JSON.stringify(location) },
+            });
+          });
+        }
+        setTimeout(() => {
```

Comment:
I am not a react expert, however a `setTimeout` that doesn't have its timeout id managed by the lifecycle of a react event seems... not good. But I assume you guys have already take this into account?

Reply:
This code isn't written in React. I'm not sure how it works. I'm guessing this is the scroll behavior after the server sends a new message.